### ISSUE TITLE
twistd should wait for the application to start when daemonizing

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11550]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| spiv reported|
|-|-|
|Trac ID|trac#823|
|Type|enhancement|
|Created|2004-12-10 21:21:52Z|
|Branch|https://github.com/twisted/twisted/tree/twistd-wait-823-5|

Currently, twistd can start successfully (exit with code 0) even if the application fails when it daemonizes. It'd be good if it waited for the application to start, returning the appropriate exit code.

<details><summary>Searchable metadata</summary>

```
trac-id__823 823
type__enhancement enhancement
reporter__spiv spiv
priority__normal normal
milestone__ 
branch__branches_twistd_wait_823_5 branches/twistd-wait-823-5
branch_author__exarkun__therve exarkun, therve
status__closed closed
resolution__fixed fixed
component__core core
keywords__twistd twistd
time__1102713712000000 1102713712000000
changetime__1374780730000000 1374780730000000
version__ 
owner__therve therve
cc__exarkun cc__spiv cc__therve cc__zooko@...
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
It would be nice if twistd had features like:
   -w --wait   twistd will wait until all services in the app configuration have
               started before exiting.  --nodaemon conflicts with this option.
   -k --kill &lt;pidfile>  stop a twisted app specified by &lt;pidfile>

--wait would be useful for scripts (e.g. functional tests) that want to spawn a
twisted app, wait for it to be listening and ready, and then proceed.  --kill
would be convenient for tearing down those tests, and probably in other places.

--wait could be implemented by opening a pipe to the child before forking, and
getting the child application to write to the pipe (or close it?) to signal when
it has started all its services.

--kill could be implemented on POSIX rather stupidly by doing:
    while True:
        try:
            os.kill(pid, 0)
        except OSError:
            break

--wait is really the one I want, though.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Shouldn't ```--wait``` just be the default?  I don't ever want to not know when the ```twistd``` command I ran didn't actually do what I wanted it to do.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(In [22480]) Branching to 'twistd-wait-823'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

(In [22481]) Throw this hack somewhere: it waits for the start of a daemonized twistd process,
and report its status.

Refs [#11550](https://github.com/twisted/twisted/issues/11550)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

(In [23470]) Branching to 'twistd-wait-823-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve commented|
|-|-|

(In [23471]) Reimplement the wait slightly differently

Refs [#11550](https://github.com/twisted/twisted/issues/11550)

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_9"></a>|@zooko commented|
|-|-|

http://tahoe-lafs.org/trac/tahoe-lafs/ticket/71 is a ticket for the Tahoe-LAFS project which I think would be fixed by using the ```--wait``` option if twistd supports that option.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_14"></a>|@zooko commented|
|-|-|

There is discussion in the Tahoe-LAFS project about whether to adapt therve's idea into Tahoe-LAFS source code base and/or to contribute such patches to this ticket: http://tahoe-lafs.org/trac/tahoe-lafs/ticket/71#comment:24

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_15"></a>|@zooko commented|
|-|-|

jml suggested on IRC that this ticket might be ready for review. I looked at [the patch](http://twistedmatrix.com/trac/changeset?new=23471%40branches/twistd-wait-823-2/twisted/scripts/_twistd_unix.py&old=23252%40trunk/twisted/scripts/_twistd_unix.py) and it is missing tests. The patch adds the functionality of having the parent wait until the child process has started and has signalled success or failure to the parent through a pipe. To test that added functionality, we might extend [twisted/test/test_twistd.py?rev=29798#L512 test_daemonize()](twisted/test/test_twistd.py?rev=29798#L512 test_daemonize()).

The tests could check whether the parent correctly exits with 0 when the child's ```startApplication()``` returns successfully, and whether the parent correctly exits with non-zero when the child's ```startApplication()``` raises an exception. I think that would be sufficient testing for this patch.

In order to do this the new ```test_daemonize()``` would have to not patch out the daemonization (as currently done in [twisted/test/test_twistd.py?rev=29798#L508 UnixApplicationRunnerSetupEnvironmentTests.setUp()](twisted/test/test_twistd.py?rev=29798#L508 UnixApplicationRunnerSetupEnvironmentTests.setUp()). Instead it could probably patch out ```os._exit()```.

I think that would be sufficient testing for this patch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [29952]) Branching to 'twistd-wait-823-3'

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_18"></a>|@therve commented|
|-|-|

(In [37525]) Branching to 'twistd-wait-823-4'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_20"></a>|@therve removed owner|
|-|-|

Let's do this!

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_21"></a>|@jerub set owner to @therve|
|-|-|

Some small comments:

The soup nazi seinfeld sketch is no longer funny. Please update this message to be something like "Daemonization tests are only run on unix platforms when  twisted.scripts._twistd_unix is available"
```
if _twistd_unix is None:
    DaemonizeTests.skip = "No daemonize for you"
```

The alignment in protocols/test/test_basic.py is wrong for these lines:
```
                          [b'produce', b'hello world', b'unproduce',
                            b'goodbye'])
```
Please fix.

I really don't like flags with names that start with 'no', and the not options['nowait'] parameter illustrates why. Can you think of a better name for this option so that we don't have to employ double negatives?

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_22"></a>|@jerub commented|
|-|-|

Okay, I did a poor review: the code I looked at was a bad diff.

Test run is available here.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_23"></a>|@therve removed owner|
|-|-|

OK, back to review. Unfortunately I haven't found a better name than `no-wait`, still opened to suggestions though.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_24"></a>|@glyph commented|
|-|-|

Replying to [jerub](#note_22):
> Test run is available here.

FYI: "here" isn't a link.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_25"></a>|@glyph commented|
|-|-|

Replying to [therve](#note_23):
> OK, back to review. Unfortunately I haven't found a better name than `no-wait`, still opened to suggestions though.

`--rush`?  `--hurry`?  :-)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_26"></a>|@glyph commented|
|-|-|

[Build results.](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/twistd-wait-823-4)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_27"></a>|@tomprince commented|
|-|-|

If I were to use this in an application, I'd ideally like
 1. any logs generated while starting up.
 2. an application hook to indicate when startup is complete. Since `.startService` isn't asynchronous, often an application won't be ready by the time it returns.

I've got some hackish code for buildbot that does it [here](https://github.com/tomprince/buildbot/compare/buildbot:master...log-magic).

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_28"></a>|@radix commented|
|-|-|


Questions:

 1. Can you explain why --no-wait is necessary? I'd rather not have it unless there's a really compelling reason.
 2. Why limit the message to 100 bytes?

Code changes:

 3. You removed the 'os' parameter from daemonize, but the docstring still documents it.
 4. There are some conflicts with trunk


Discussion:

 5. I'm not super into monkeypatch-the-import; it's pretty sensitive to changes. But I guess we already do it elsewhere.
 5. the move of the daemonize method was hard to review.
 7. I notice that this branch doesn't implement --kill, but it's clear to me that --kill should be done in a separate ticket/branch, so that's cool.

Everything else about the change seems pretty reasonable, but I'm a little worried that there may be weird edge cases that this would break for. Oh well!

To address tom's point:

 1. I think logs would be pretty cool, but a notification that the user should look at the log file is pretty good for now.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_30"></a>|@therve commented|
|-|-|

Tom: your second point is interesting, but really unrelated to anything happening in that ticket. It's already 9 years old, let's not add random feature requests.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_31"></a>|@therve commented|
|-|-|

(In [38885]) Branching to 'twistd-wait-823-5'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_32"></a>|@therve removed owner|
|-|-|

Replying to [radix](#note_28):
> 
> Questions:
> 
>  1. Can you explain why --no-wait is necessary? I'd rather not have it unless there's a really compelling reason.

The idea was that people may want to fall back to previous behavior. Wait is the new default behavior because it's the one that makes most sense, but maybe people care about firing without caring?

>  2. Why limit the message to 100 bytes?

We need a limit, I picked an arbitrary number.

> Code changes:
> 
>  3. You removed the 'os' parameter from daemonize, but the docstring still documents it.
>  4. There are some conflicts with trunk

Fixed.
 
> Discussion:
> 
>  5. I'm not super into monkeypatch-the-import; it's pretty sensitive to changes. But I guess we already do it elsewhere.

Yeah it's not super great.

>  5. the move of the daemonize method was hard to review.

Sorry, http://paste.pound-python.org/show/34194/ FWIW.

>  7. I notice that this branch doesn't implement --kill, but it's clear to me that --kill should be done in a separate ticket/branch, so that's cool.

Right we can open a bug for that.

> Everything else about the change seems pretty reasonable, but I'm a little worried that there may be weird edge cases that this would break for. Oh well!

Maybe! As seen on the previous diff, the actual changes are fairly minimal though.


Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_33"></a>|@glyph commented|
|-|-|

Replying to [therve](#note_32):
> Replying to [radix](#note_28):
> > 
> > Questions:
> > 
> >  1. Can you explain why --no-wait is necessary? I'd rather not have it unless there's a really compelling reason.
> 
> The idea was that people may want to fall back to previous behavior. Wait is the new default behavior because it's the one that makes most sense, but maybe people care about firing without caring?

Can't the user just do this with "`nohup twistd ... &`" on UNIXish platforms (and by switching to a UNIXish platform if they're on Windows? :)).

> >  7. I notice that this branch doesn't implement --kill, but it's clear to me that --kill should be done in a separate ticket/branch, so that's cool.
> 
> Right we can open a bug for that.

Please do.  And link to it ;).

> > Everything else about the change seems pretty reasonable, but I'm a little worried that there may be weird edge cases that this would break for. Oh well!
> 
> Maybe! As seen on the previous diff, the actual changes are fairly minimal though.

That's what the pre-release period is for, right? :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_34"></a>|@therve commented|
|-|-|

Replying to [glyph](#note_33):
> Replying to [therve](#note_32):
> > Replying to [radix](#note_28):
> > > 
> > > Questions:
> > > 
> > >  1. Can you explain why --no-wait is necessary? I'd rather not have it unless there's a really compelling reason.
> > 
> > The idea was that people may want to fall back to previous behavior. Wait is the new default behavior because it's the one that makes most sense, but maybe people care about firing without caring?
> 
> Can't the user just do this with "`nohup twistd ... &`" on UNIXish platforms (and by switching to a UNIXish platform if they're on Windows? :)).

They sure can. Windows is irrelevant here because it's a daemonization feature. Should I remove the argument then?

> > >  7. I notice that this branch doesn't implement --kill, but it's clear to me that --kill should be done in a separate ticket/branch, so that's cool.
> > 
> > Right we can open a bug for that.
> 
> Please do.  And link to it ;).

[#6593](https://github.com/twisted/twisted/issues/6593).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_35"></a>|@glyph commented|
|-|-|

Replying to [therve](#note_34):
> They sure can. Windows is irrelevant here because it's a daemonization feature. Should I remove the argument then?

I honestly can't think of a good reason to keep it, and all things being equal, we should not have more stuff.  Please remove.

> [#6593](https://github.com/twisted/twisted/issues/6593).

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_37"></a>|@therve removed owner|
|-|-|

Back for review! I removed the flag, and also added some untilConcludes calls for safety.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_38"></a>|@wallrj set owner to @therve|
|-|-|

## Code Review

Thanks Therve, this looks like a great improvement.

Notes:
 * Merges cleanly
 * trial twisted.test passes on Fedora 19
 * Attempted to run twistd dns on privilieged port which resulted in
   the expected error. (although I wonder if the stale pid message
   should be in the log file rather than stderr?)
   ```
   [richard@zorin twistd-wait-823-5]$ ./bin/twistd dns
   Removing stale pidfile /home/richard/projects/Twisted/branches/twistd-wait-823-5/twistd.pid
   An error has occurred: 'Couldn't listen on any:53: [Errno 13] Permission denied.'
   Please look at log file for more information.
   ```


Some points:
 1. Build failures: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/twistd-wait-823-5
    1. winxp32-py2.7: https://buildbot.twistedmatrix.com/builders/winxp32-py2.7/builds/2558/steps/select/logs/problems
       Could this be caused by your changes to test_process? Or is it just a routine windows failure?
       ```
       [ERROR]
       Traceback (most recent call last):
         File "C:\buildslave\twisted\winxp32-py2.7\Twisted\twisted\internet\base.py", line 824, in runUntilCurrent
           call.func(*call.args, **call.kw)
         File "C:\buildslave\twisted\winxp32-py2.7\Twisted\twisted\internet\_pollingfile.py", line 79, in _pollEvent
           workUnits += resource.checkWork()
         File "C:\buildslave\twisted\winxp32-py2.7\Twisted\twisted\internet\_pollingfile.py", line 136, in checkWork
           self.cleanup()
         File "C:\buildslave\twisted\winxp32-py2.7\Twisted\twisted\internet\_pollingfile.py", line 141, in cleanup
           self.lostCallback()
         File "C:\buildslave\twisted\winxp32-py2.7\Twisted\twisted\internet\_dumbwin32proc.py", line 336, in outConnectionLost
           self.proto.childConnectionLost(1)
         File "C:\buildslave\twisted\winxp32-py2.7\Twisted\twisted\test\test_process.py", line 204, in childConnectionLost
           "Data was %r instead of 'abcd'" % (self.data,))
       exceptions.RuntimeError: Data was '' instead of 'abcd'

       twisted.test.test_process.ProcessTestCase.testManyProcesses
       ```
    2. twistedchecker: https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/1043/steps/run-twistedchecker/logs/new%20twistedchecker%20errors
       ```
       ************* Module twisted.scripts._twistd_unix
       W9604:308,12:UnixApplicationRunner.daemonize: Please use 'except Exception as e:',rather than 'except Exception, e:'
       ************* Module twisted.test.test_process
       W9202:1382,4:MockOS.read: Missing epytext markup @param for argument "fd"
       W9202:1382,4:MockOS.read: Missing epytext markup @param for argument "size"
       W9204:1382,4:MockOS.read: Missing epytext markup @return for return value
       W9202:1516,4:MockOS.chdir: Missing epytext markup @param for argument "path"
       W9202:1582,4:MockOS.unlink: Missing epytext markup @param for argument "filename"
       W9202:1589,4:MockOS.umask: Missing epytext markup @param for argument "mask"
       W9204:1596,4:MockOS.getpid: Missing epytext markup @return for return value
       C0301:1760,0: Line too long (86/79)
       ```
    3. pyflakes: https://buildbot.twistedmatrix.com/builders/pyflakes/builds/754/steps/pyflakes/logs/new%20pyflakes%20errors
       ```
       twisted/test/test_twistd.py:19: redefinition of unused 'grp' from line 17
       twisted/test/test_twistd.py:19: redefinition of unused 'pwd' from line 16
       twisted/test/test_twistd.py:24: redefinition of unused 'pickle' from line 22
       twisted/test/test_twistd.py:46: redefinition of unused '_twistd_unix' from line 44
       twisted/test/test_twistd.py:55: redefinition of unused 'syslog' from line 53
       twisted/test/test_twistd.py:61: redefinition of unused 'profile' from line 59
       twisted/test/test_twistd.py:70: redefinition of unused 'hotshot' from line 65
       twisted/test/test_twistd.py:76: redefinition of unused 'cProfile' from line 74
       ```
    4. coverage: Some parts of daemonize aren't covered. They never
       were, but you may want to improve
       things.
       * https://buildbot.twistedmatrix.com/builds/twisted-coverage.py/twisted-coverage.py-r38947/twisted_scripts__twistd_unix.html#n275

 2. [branches/twistd-wait-823-5/twisted/scripts/_twistd_unix.py](branches/twistd-wait-823-5/twisted/scripts/_twistd_unix.py)
    1. "code = 0": That assigment looks redundant
       ```
       298              if os.fork():  # launch child and...
               299                  code = 0
               300                  code = self._waitForStart(r)
       ```

    2. "Please look at log file for more information": It would be nice
       to print the log file path here (self.config['logfile']) but
       only if it's easy.

    3. I wondered why you used "sys.__stderr__" rather than
       "sys.stderr"?

    4. The MockOS tests are clever and fast, but I'd have a bit more
       confidence if there were some functional tests to launch a
       twistd process and check its exit status. Maybe that's
       something for another ticket.

This branch has already gone through 4 reviews and the feature seems
to work well, so I'm happy for you to merge after addressing or
answering the points above and checking for some clean build results.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_39"></a>|@therve set status to `closed`|
|-|-|

(In [39314]) Merge twistd-wait-823-5

Author: therve
Reviewers: rwall, radix, tom.prince, jerub
Fixes: [#11550](https://github.com/twisted/twisted/issues/11550)

twistd now waits for the application to start successfully before exiting after
daemonization.
